File: src/config/ship-groups.ts
import rawShipGroups from '../../schemas/ship-groups.json' with { type: 'json' };
import { log } from '../utils/log.js';
import type { CanonicalVariantCode } from '../lib/canon.js';

interface RawHullConfig {
  manufacturer?: unknown;
  name?: unknown;
  variants?: Record<string, unknown>;
}

interface VariantConfiguration {
  code: string;
  match: string;
}

interface NormalizedVariantConfig {
  matchIds: string[];
  shipVariantIds: string[];
  variantCodes: CanonicalVariantCode[];
  names: string[];
  editions: string[];
  configurations: VariantConfiguration[];
}

interface HullDefinition {
  hullKey: string;
  manufacturer: string;
  name: string;
}

export interface VariantAssignment extends HullDefinition {
  variantCode: CanonicalVariantCode;
  names: string[];
  editions: string[];
  configurations: VariantConfiguration[];
}

function toString(value: unknown): string | undefined {
  if (typeof value === 'string') {
    const trimmed = value.trim();
    return trimmed ? trimmed : undefined;
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  return undefined;
}

function collectStrings(value: unknown): string[] {
  if (value === undefined || value === null) return [];
  const source = Array.isArray(value) ? value : [value];
  const result: string[] = [];
  for (const entry of source) {
    const str = toString(entry);
    if (!str) continue;
    result.push(str);
  }
  return result;
}

function dedupeStrings(values: string[]): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const value of values) {
    const trimmed = value.trim();
    if (!trimmed) continue;
    const key = trimmed.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    result.push(trimmed);
  }
  return result;
}

function sanitizeToken(value: string | undefined): string | undefined {
  if (!value) return undefined;
  const normalized = value
    .replace(/[^a-z0-9]+/gi, '_')
    .replace(/_+/g, '_')
    .replace(/^_|_$/g, '')
    .toUpperCase();
  return normalized || undefined;
}

function sanitizeHullKey(value: string | undefined): string {
  return sanitizeToken(value) ?? 'HULL';
}

function sanitizeVariantCode(value: string | undefined): CanonicalVariantCode {
  return sanitizeToken(value) ?? 'BASE';
}

function normalizeVariant(input: unknown): NormalizedVariantConfig {
  if (typeof input === 'string' || Array.isArray(input)) {
    const matchIds = dedupeStrings(collectStrings(input));
    return {
      matchIds,
      shipVariantIds: [],
      variantCodes: [],
      names: [],
      editions: [],
      configurations: []
    };
  }

  if (!input || typeof input !== 'object') {
    return { matchIds: [], shipVariantIds: [], variantCodes: [], names: [], editions: [], configurations: [] };
  }

  const record = input as Record<string, unknown>;
  const matchIds = dedupeStrings([
    ...collectStrings(record.match ?? (record as any).match_id),
    ...collectStrings(record.id),
    ...collectStrings(record.ids),
    ...collectStrings(record.matches)
  ]);

  const shipVariantIds = dedupeStrings(
    collectStrings(record.ship_variant_ids ?? record.shipVariantIds)
  );
  const variantCodes = dedupeStrings(
    collectStrings(
      record.ship_variant_codes ??
        record.shipVariantCodes ??
        record.codes ??
        record.variant_codes ??
        record.aliases
    )
  )
    .map((code) => sanitizeVariantCode(code))
    .filter((code, index, array) => array.indexOf(code) === index);
  const names = dedupeStrings(collectStrings(record.names ?? record.display_names));
  const editions = dedupeStrings(collectStrings(record.editions ?? record.profiles));

  const configurations: VariantConfiguration[] = [];
  const rawConfigurations = record.configurations;
  if (rawConfigurations && typeof rawConfigurations === 'object' && !Array.isArray(rawConfigurations)) {
    for (const [rawCode, rawMatch] of Object.entries(rawConfigurations as Record<string, unknown>)) {
      const match = toString(rawMatch);
      if (!match) continue;
      const code = sanitizeToken(rawCode) ?? sanitizeToken(match);
      if (!code) continue;
      configurations.push({ code, match });
    }
  }

  if (!matchIds.length && configurations.length) {
    const fallbackMatch = configurations[0]?.match;
    if (fallbackMatch) {
      matchIds.push(fallbackMatch);
    }
  }

  return { matchIds, shipVariantIds, variantCodes, names, editions, configurations };
}

function normalizeLookupKey(value: string | undefined): string | undefined {
  if (!value) return undefined;
  const trimmed = value.trim();
  return trimmed ? trimmed.toLowerCase() : undefined;
}

function assignmentsEqual(a: VariantAssignment, b: VariantAssignment): boolean {
  return a.hullKey === b.hullKey && a.variantCode === b.variantCode;
}

export class ShipGrouping {
  private readonly hulls = new Map<string, HullDefinition>();
  private readonly canonicalVariants = new Map<string, VariantAssignment>();
  private readonly perHullVariants = new Map<string, Map<string, VariantAssignment>>();
  private readonly shipIdAssignments = new Map<string, VariantAssignment>();
  private readonly shipVariantIdAssignments = new Map<string, VariantAssignment>();

  constructor(rawConfig: unknown = rawShipGroups) {
    const root = rawConfig && typeof rawConfig === 'object' ? (rawConfig as Record<string, unknown>) : {};
    const rawHulls = root.hulls;

    if (rawHulls && typeof rawHulls === 'object') {
      for (const [rawHullKey, rawHullValue] of Object.entries(rawHulls as Record<string, RawHullConfig>)) {
        const hullKey = sanitizeHullKey(rawHullKey);
        const manufacturer = toString(rawHullValue?.manufacturer) ?? hullKey.split('_')[0] ?? 'UNKNOWN';
        const name = toString(rawHullValue?.name) ?? hullKey;
        const hullDef: HullDefinition = { hullKey, manufacturer, name };
        this.hulls.set(hullKey, hullDef);

        const variantLookup = new Map<string, VariantAssignment>();
        this.perHullVariants.set(hullKey, variantLookup);

        const rawVariants = rawHullValue?.variants;
        if (!rawVariants || typeof rawVariants !== 'object') continue;

        for (const [rawVariantKey, variantValue] of Object.entries(rawVariants)) {
          const variantCode = sanitizeVariantCode(rawVariantKey);
          const normalized = normalizeVariant(variantValue);
          const assignment: VariantAssignment = {
            ...hullDef,
            variantCode,
            names: normalized.names,
            editions: normalized.editions,
            configurations: [...normalized.configurations]
          };

          this.registerVariant(hullKey, assignment, normalized);
        }
      }
    }
  }

  private registerVariant(
    hullKey: string,
    assignment: VariantAssignment,
    config: NormalizedVariantConfig
  ): void {
    const canonicalKey = `${hullKey}:${assignment.variantCode}`;
    const existingCanonical = this.canonicalVariants.get(canonicalKey);
    if (existingCanonical && !assignmentsEqual(existingCanonical, assignment)) {
      log.warn('Ship grouping config collision on canonical variant', {
        hull: hullKey,
        variant: assignment.variantCode
      });
      return;
    }
    this.canonicalVariants.set(canonicalKey, assignment);

    const variantLookup = this.perHullVariants.get(hullKey);
    if (!variantLookup) return;

    const variantCodes = new Set<CanonicalVariantCode>([assignment.variantCode]);
    for (const alias of config.variantCodes) {
      variantCodes.add(alias);
    }

    for (const code of variantCodes) {
      const existing = variantLookup.get(code);
      if (existing && !assignmentsEqual(existing, assignment)) {
        log.warn('Ship grouping variant code collision', {
          hull: hullKey,
          variant: assignment.variantCode,
          code
        });
        continue;
      }
      variantLookup.set(code, assignment);
    }

    const matchIds = config.matchIds.length ? config.matchIds : [assignment.variantCode];
    for (const id of matchIds) {
      this.registerShipId(id, assignment);
    }

    for (const id of config.shipVariantIds) {
      this.registerShipVariantId(id, assignment);
    }
  }

  private registerShipId(id: string, assignment: VariantAssignment): void {
    const key = normalizeLookupKey(id);
    if (!key) return;
    const existing = this.shipIdAssignments.get(key);
    if (existing && !assignmentsEqual(existing, assignment)) {
      log.warn('Ship grouping ship identifier collision', {
        id,
        hull: assignment.hullKey,
        variant: assignment.variantCode
      });
      return;
    }
    this.shipIdAssignments.set(key, assignment);
  }

  private registerShipVariantId(id: string, assignment: VariantAssignment): void {
    const key = normalizeLookupKey(id);
    if (!key) return;
    const existing = this.shipVariantIdAssignments.get(key);
    if (existing && !assignmentsEqual(existing, assignment)) {
      log.warn('Ship grouping ship_variant identifier collision', {
        id,
        hull: assignment.hullKey,
        variant: assignment.variantCode
      });
      return;
    }
    this.shipVariantIdAssignments.set(key, assignment);
  }

  getHull(hullKey: string): HullDefinition | undefined {
    return this.hulls.get(sanitizeHullKey(hullKey));
  }

  lookupShipId(...candidates: (string | undefined)[]): VariantAssignment | undefined {
    for (const candidate of candidates) {
      const key = normalizeLookupKey(candidate);
      if (!key) continue;
      const assignment = this.shipIdAssignments.get(key);
      if (assignment) {
        return assignment;
      }
    }
    return undefined;
  }

  lookupShipVariantId(id: string | undefined): VariantAssignment | undefined {
    const key = normalizeLookupKey(id);
    if (!key) return undefined;
    return this.shipVariantIdAssignments.get(key);
  }

  lookupVariant(hullKey: string, ...candidateCodes: (string | undefined)[]): VariantAssignment | undefined {
    const lookup = this.perHullVariants.get(sanitizeHullKey(hullKey));
    if (!lookup) return undefined;

    for (const candidate of candidateCodes) {
      const normalized = candidate ? sanitizeVariantCode(candidate) : undefined;
      if (!normalized) continue;
      const assignment = lookup.get(normalized);
      if (assignment) return assignment;
    }

    return undefined;
  }

  entries(): VariantAssignment[] {
    return [...this.canonicalVariants.values()];
  }

  resolveConfiguration(assignment: VariantAssignment, ...candidates: (string | undefined)[]): string | undefined {
    if (!assignment.configurations.length) return undefined;
    const normalizedCandidates = candidates
      .map((candidate) => (candidate ? candidate.toLowerCase() : undefined))
      .filter((candidate): candidate is string => Boolean(candidate));
    if (!normalizedCandidates.length) return undefined;

    for (const configuration of assignment.configurations) {
      const normalizedTarget = configuration.match.trim().toLowerCase();
      if (!normalizedTarget) continue;
      for (const candidate of normalizedCandidates) {
        if (candidate.includes(normalizedTarget)) {
          return configuration.code;
        }
      }
    }

    return undefined;
  }
}

export function loadShipGrouping(): ShipGrouping {
  return new ShipGrouping();
}

File: internal/transform/shipgroups.go
package transform

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/ArisCorporation/sc-goetl/internal/lib"
	"github.com/ArisCorporation/sc-goetl/internal/utils"
)

// VariantAssignment mirrors the TypeScript structure.
type VariantAssignment struct {
	HullKey        string
	Manufacturer   string
	Name           string
	VariantCode    lib.CanonicalVariantCode
	Names          []string
	Editions       []string
	MatchIDs       []string
	ShipVariantIDs []string
	Configurations []VariantConfiguration
}

// VariantConfiguration stores per-hull configuration metadata.
type VariantConfiguration struct {
	Code           string
	Name           string
	Match          string
	ShipVariantIDs []string
	Profiles       []string
}

// ShipGrouping provides lookup utilities for variant metadata.
type ShipGrouping struct {
	hulls                  map[string]HullDefinition
	canonicalVariants      map[string]VariantAssignment
	perHullVariants        map[string]map[string]VariantAssignment
	shipIDAssignments      map[string]VariantAssignment
	shipVariantAssignments map[string]VariantAssignment
}

func emptyShipGrouping() *ShipGrouping {
	return &ShipGrouping{
		hulls:                  map[string]HullDefinition{},
		canonicalVariants:      map[string]VariantAssignment{},
		perHullVariants:        map[string]map[string]VariantAssignment{},
		shipIDAssignments:      map[string]VariantAssignment{},
		shipVariantAssignments: map[string]VariantAssignment{},
	}
}

// HullDefinition describes a hull entry.
type HullDefinition struct {
	HullKey      string
	Manufacturer string
	Name         string
}

// LoadShipGrouping loads configuration from schemas/ship-groups.json.
func LoadShipGrouping() *ShipGrouping {
	path := filepath.Join("schemas", "ship-groups.json")
	bytes, err := os.ReadFile(path)
	if err != nil {
		utils.Logger().Warn("Ship grouping configuration missing", "error", err)
		return emptyShipGrouping()
	}
	var raw map[string]any
	if err := json.Unmarshal(bytes, &raw); err != nil {
		utils.Logger().Warn("Failed to parse ship grouping configuration", "error", err)
		return emptyShipGrouping()
	}
	grouping := &ShipGrouping{
		hulls:                  map[string]HullDefinition{},
		canonicalVariants:      map[string]VariantAssignment{},
		perHullVariants:        map[string]map[string]VariantAssignment{},
		shipIDAssignments:      map[string]VariantAssignment{},
		shipVariantAssignments: map[string]VariantAssignment{},
	}
	rawHulls, _ := raw["hulls"].(map[string]any)
	for rawHullKey, rawHullValue := range rawHulls {
		hullKey := sanitizeHullKey(rawHullKey)
		rawConfig, _ := rawHullValue.(map[string]any)
		manufacturer := toString(rawConfig["manufacturer"])
		if manufacturer == "" {
			parts := strings.SplitN(hullKey, "_", 2)
			if len(parts) > 0 {
				manufacturer = parts[0]
			} else {
				manufacturer = "UNKNOWN"
			}
		}
		name := toString(rawConfig["name"])
		if name == "" {
			name = hullKey
		}
		hullDef := HullDefinition{
			HullKey:      hullKey,
			Manufacturer: manufacturer,
			Name:         name,
		}
		grouping.hulls[hullKey] = hullDef
		grouping.perHullVariants[hullKey] = map[string]VariantAssignment{}

		rawVariants, _ := rawConfig["variants"].(map[string]any)
		for rawVariantKey, rawVariantValue := range rawVariants {
			variantCode := sanitizeVariantCode(rawVariantKey)
			normalized := normalizeVariant(rawVariantValue)
			assignment := VariantAssignment{
				HullKey:        hullDef.HullKey,
				Manufacturer:   hullDef.Manufacturer,
				Name:           hullDef.Name,
				VariantCode:    variantCode,
				Names:          normalized.Names,
				Editions:       normalized.Editions,
				MatchIDs:       normalized.MatchIDs,
				ShipVariantIDs: normalized.ShipVariantIDs,
				Configurations: normalized.Configurations,
			}
			grouping.registerVariant(hullKey, assignment, normalized)
		}
	}
	return grouping
}

func sanitizeHullKey(value string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return "HULL"
	}
	return sanitizeToken(value)
}

func sanitizeVariantCode(value string) lib.CanonicalVariantCode {
	value = strings.TrimSpace(value)
	if value == "" {
		return "BASE"
	}
	return lib.CanonicalVariantCode(sanitizeToken(value))
}

func sanitizeToken(value string) string {
	value = strings.ToUpper(value)
	builder := strings.Builder{}
	lastUnderscore := false
	for _, r := range value {
		if (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			builder.WriteRune(r)
			lastUnderscore = false
			continue
		}
		if !lastUnderscore {
			builder.WriteRune('_')
			lastUnderscore = true
		}
	}
	result := builder.String()
	result = strings.Trim(result, "_")
	result = strings.ReplaceAll(result, "__", "_")
	return result
}

func toString(value any) string {
	switch v := value.(type) {
	case string:
		return strings.TrimSpace(v)
	case fmt.Stringer:
		return strings.TrimSpace(v.String())
	case float64:
		return strings.TrimSpace(fmt.Sprintf("%.0f", v))
	case int, int64:
		return strings.TrimSpace(fmt.Sprint(v))
	default:
		return ""
	}
}

type normalizedVariantConfig struct {
	MatchIDs       []string
	ShipVariantIDs []string
	VariantCodes   []lib.CanonicalVariantCode
	Names          []string
	Editions       []string
	Configurations []VariantConfiguration
}

func normalizeVariant(input any) normalizedVariantConfig {
	switch value := input.(type) {
	case string:
		if value == "" {
			return normalizedVariantConfig{}
		}
		return normalizedVariantConfig{
			MatchIDs: []string{value},
		}
	case []any:
		match := []string{}
		for _, entry := range value {
			if s := toString(entry); s != "" {
				match = append(match, s)
			}
		}
		return normalizedVariantConfig{
			MatchIDs: match,
		}
	case map[string]any:
		matchIDs := collectStrings(value["match"], value["match_id"], value["id"], value["ids"], value["matches"])
		if len(matchIDs) == 0 {
			if configs, ok := value["configurations"].(map[string]any); ok && len(configs) > 0 {
				for _, configValue := range configs {
					if entry, ok := configValue.(map[string]any); ok {
						if match := toString(entry["match"]); match != "" {
							matchIDs = append(matchIDs, match)
						}
					}
				}
			}
		}
		shipVariantIDs := collectStrings(value["ship_variant_ids"], value["shipVariantIds"])
		variantCodes := []lib.CanonicalVariantCode{}
		for _, raw := range collectStrings(value["ship_variant_codes"], value["shipVariantCodes"], value["codes"], value["variant_codes"], value["aliases"]) {
			variantCodes = append(variantCodes, sanitizeVariantCode(raw))
		}
		names := collectStrings(value["names"], value["display_names"])
		editions := collectStrings(value["editions"], value["profiles"])

		configs := []VariantConfiguration{}
		if rawConfigs, ok := value["configurations"].(map[string]any); ok {
			for rawCode, configValue := range rawConfigs {
				code := sanitizeToken(rawCode)
				if code == "" {
					code = sanitizeToken(toString(configValue))
				}
				if code == "" {
					continue
				}
				if entry, ok := configValue.(map[string]any); ok {
					name := toString(entry["name"])
					if name == "" {
						name = toString(entry["label"])
					}
					if name == "" {
						name = toString(entry["display"])
					}
					if name == "" {
						name = rawCode
					}
					match := toString(entry["match"])
					if match == "" {
						match = toString(entry["class"])
					}
					if match == "" {
						match = toString(entry["id"])
					}
					configShipVariantIDs := collectStrings(entry["ship_variant_ids"], entry["shipVariantIds"], entry["ids"])
					profiles := collectStrings(entry["profiles"], entry["profile"])
					configs = append(configs, VariantConfiguration{
						Code:           code,
						Name:           name,
						Match:          match,
						ShipVariantIDs: configShipVariantIDs,
						Profiles:       profiles,
					})
				} else {
					match := toString(configValue)
					if match == "" {
						match = rawCode
					}
					configs = append(configs, VariantConfiguration{
						Code:  code,
						Name:  match,
						Match: match,
					})
				}
			}
		}
		return normalizedVariantConfig{
			MatchIDs:       dedupeStrings(matchIDs),
			ShipVariantIDs: dedupeStrings(shipVariantIDs),
			VariantCodes:   variantCodes,
			Names:          dedupeStrings(names),
			Editions:       dedupeStrings(editions),
			Configurations: configs,
		}
	default:
		return normalizedVariantConfig{}
	}
}

func collectStrings(values ...any) []string {
	set := []string{}
	for _, value := range values {
		switch v := value.(type) {
		case string:
			if s := strings.TrimSpace(v); s != "" {
				set = append(set, s)
			}
		case []any:
			for _, entry := range v {
				if s := toString(entry); s != "" {
					set = append(set, s)
				}
			}
		case []string:
			for _, entry := range v {
				if s := strings.TrimSpace(entry); s != "" {
					set = append(set, s)
				}
			}
		}
	}
	return set
}

func normalizeLookupKey(value string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}
	return strings.ToLower(value)
}

func (g *ShipGrouping) registerVariant(hullKey string, assignment VariantAssignment, config normalizedVariantConfig) {
	canonicalKey := hullKey + ":" + string(assignment.VariantCode)
	if existing, ok := g.canonicalVariants[canonicalKey]; ok && !assignmentsEqual(existing, assignment) {
		utils.Logger().Warn("Ship grouping config collision on canonical variant", "hull", hullKey, "variant", assignment.VariantCode)
		return
	}
	g.canonicalVariants[canonicalKey] = assignment

	variantLookup := g.perHullVariants[hullKey]
	if variantLookup == nil {
		variantLookup = map[string]VariantAssignment{}
		g.perHullVariants[hullKey] = variantLookup
	}

	variantCodes := map[lib.CanonicalVariantCode]struct{}{
		assignment.VariantCode: {},
	}
	for _, alias := range config.VariantCodes {
		variantCodes[alias] = struct{}{}
	}
	for code := range variantCodes {
		if existing, ok := variantLookup[string(code)]; ok && !assignmentsEqual(existing, assignment) {
			utils.Logger().Warn("Ship grouping variant code collision", "hull", hullKey, "variant", assignment.VariantCode, "code", code)
			continue
		}
		variantLookup[string(code)] = assignment
	}

	matchIDs := config.MatchIDs
	if len(matchIDs) == 0 {
		matchIDs = []string{string(assignment.VariantCode)}
	}
	for _, id := range matchIDs {
		g.registerShipID(id, assignment)
	}
	for _, id := range config.ShipVariantIDs {
		g.registerShipVariantID(id, assignment)
	}
	for _, id := range assignment.ShipVariantIDs {
		g.registerShipVariantID(id, assignment)
	}
	for _, variantConfig := range assignment.Configurations {
		for _, id := range variantConfig.ShipVariantIDs {
			g.registerShipVariantID(id, assignment)
		}
	}
}

func (g *ShipGrouping) registerShipID(id string, assignment VariantAssignment) {
	key := normalizeLookupKey(id)
	if key == "" {
		return
	}
	if existing, ok := g.shipIDAssignments[key]; ok && !assignmentsEqual(existing, assignment) {
		utils.Logger().Warn("Ship grouping ship identifier collision", "id", id, "hull", assignment.HullKey, "variant", assignment.VariantCode)
		return
	}
	g.shipIDAssignments[key] = assignment
}

func (g *ShipGrouping) registerShipVariantID(id string, assignment VariantAssignment) {
	key := normalizeLookupKey(id)
	if key == "" {
		return
	}
	if existing, ok := g.shipVariantAssignments[key]; ok && !assignmentsEqual(existing, assignment) {
		utils.Logger().Warn("Ship grouping ship_variant identifier collision", "id", id, "hull", assignment.HullKey, "variant", assignment.VariantCode)
		return
	}
	g.shipVariantAssignments[key] = assignment
}

func assignmentsEqual(a, b VariantAssignment) bool {
	return a.HullKey == b.HullKey && a.VariantCode == b.VariantCode
}

// GetHull returns a hull definition.
func (g *ShipGrouping) GetHull(hullKey string) (HullDefinition, bool) {
	def, ok := g.hulls[sanitizeHullKey(hullKey)]
	return def, ok
}

// LookupShipID resolves a ship by id candidates.
func (g *ShipGrouping) LookupShipID(candidates ...string) (VariantAssignment, bool) {
	for _, candidate := range candidates {
		key := normalizeLookupKey(candidate)
		if key == "" {
			continue
		}
		if assignment, ok := g.shipIDAssignments[key]; ok {
			return assignment, true
		}
	}
	return VariantAssignment{}, false
}

// LookupShipVariantID resolves a variant by identifier.
func (g *ShipGrouping) LookupShipVariantID(id string) (VariantAssignment, bool) {
	key := normalizeLookupKey(id)
	if key == "" {
		return VariantAssignment{}, false
	}
	assignment, ok := g.shipVariantAssignments[key]
	return assignment, ok
}

// LookupVariant resolves a variant for the given hull and candidate codes.
func (g *ShipGrouping) LookupVariant(hullKey string, candidateCodes ...string) (VariantAssignment, bool) {
	lookup := g.perHullVariants[sanitizeHullKey(hullKey)]
	if lookup == nil {
		return VariantAssignment{}, false
	}
	for _, candidate := range candidateCodes {
		normalized := sanitizeVariantCode(candidate)
		if assignment, ok := lookup[string(normalized)]; ok {
			return assignment, true
		}
	}
	return VariantAssignment{}, false
}

// Entries exposes canonical assignments.
func (g *ShipGrouping) Entries() []VariantAssignment {
	result := make([]VariantAssignment, 0, len(g.canonicalVariants))
	for _, assignment := range g.canonicalVariants {
		result = append(result, assignment)
	}
	return result
}

// ResolveConfiguration finds a configuration code for provided candidates.
func (g *ShipGrouping) ResolveConfiguration(assignment VariantAssignment, candidates ...string) string {
	if len(assignment.Configurations) == 0 {
		return ""
	}
	normalized := []string{}
	for _, candidate := range candidates {
		candidate = strings.TrimSpace(strings.ToLower(candidate))
		if candidate != "" {
			normalized = append(normalized, candidate)
		}
	}
	if len(normalized) == 0 {
		return ""
	}
	for _, configuration := range assignment.Configurations {
		targets := map[string]struct{}{}
		addTarget := func(value string) {
			if value == "" {
				return
			}
			targets[strings.TrimSpace(strings.ToLower(value))] = struct{}{}
		}
		addTarget(configuration.Match)
		addTarget(configuration.Name)
		addTarget(configuration.Code)
		for _, profile := range configuration.Profiles {
			addTarget(profile)
		}
		for _, variantID := range configuration.ShipVariantIDs {
			addTarget(variantID)
		}
		for _, candidate := range normalized {
			if _, ok := targets[candidate]; ok {
				return configuration.Code
			}
		}
	}
	return ""
}
